src/main/java/org/folio/rest/impl/ProtectionHelper.java

```diff
@@ -0,0 +1,142 @@
+package org.folio.rest.impl;
+
+import io.vertx.core.Context;
+import org.folio.HttpStatus;
+import org.folio.orders.rest.exceptions.HttpException;
+import org.folio.orders.utils.ProtectedOperationType;
+import org.folio.rest.jaxrs.model.*;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.stream.Collectors.toList;
+import static org.folio.orders.utils.ErrorCodes.*;
+import static org.folio.orders.utils.HelperUtils.convertIdsToCqlQuery;
+import static org.folio.rest.RestVerticle.OKAPI_USERID_HEADER;
+
+public class ProtectionHelper extends AbstractHelper {
+
+  private AcquisitionsUnitsHelper acquisitionsUnitsHelper;
+  private AcquisitionsUnitAssignmentsHelper acquisitionsUnitAssignmentsHelper;
+  private ProtectedOperationType operation;
+
+
+  public ProtectionHelper(Map<String, String> okapiHeaders, Context ctx, String lang, ProtectedOperationType operation) {
+    super(okapiHeaders, ctx, lang);
+    acquisitionsUnitsHelper = new AcquisitionsUnitsHelper(okapiHeaders, ctx, lang);
+    acquisitionsUnitAssignmentsHelper = new AcquisitionsUnitAssignmentsHelper(okapiHeaders, ctx, lang);
+    this.operation = operation;
+  }
+
+  /**
+   * This method determines status of operation restriction based on ID of {@link CompositePurchaseOrder}.
+   * @param recordId corresponding record ID.
+   *
+   * @return true if operation is restricted, otherwise - false.
+   */
+  public CompletableFuture<Boolean> isOperationRestricted(String recordId) {
+    String userId;
+    if(okapiHeaders != null && (userId = okapiHeaders.get(OKAPI_USERID_HEADER)) != null) {
+      return getUnitIdsAssignedToOrder(recordId)
+        .thenCompose(unitIds -> isOperationRestricted(userId, unitIds));
+    } else {
+      throw new HttpException(HttpStatus.HTTP_FORBIDDEN.toInt(), UNKNOWN_USER);
+    }
+  }
+
+  /**
+   * This method determines status of operation restriction based on unit IDs from {@link CompositePurchaseOrder}.
+   * @param unitIds list of unit IDs.
+   *
+   * @return true if operation is restricted, otherwise - false.
+   */
+  public CompletableFuture<Boolean> isOperationRestricted(List<String> unitIds) {
+    String userId;
+    if(okapiHeaders != null && (userId = okapiHeaders.get(OKAPI_USERID_HEADER)) != null) {
+        return isOperationRestricted(userId, unitIds);
+    } else {
+      throw new HttpException(HttpStatus.HTTP_FORBIDDEN.toInt(), UNKNOWN_USER);
+    }
+  }
+
+  /**
+   * This method determines status of operation restriction based on units analysis and user ID.
+   * @param unitIds list of unit's IDs.
+   * @param userId user's ID.
+   *
+   * @return true if operation is restricted, otherwise - false.
+   */
+  private CompletableFuture<Boolean> isOperationRestricted(String userId, List<String> unitIds) {
+    if(!unitIds.isEmpty()) {
+      CompletableFuture<Boolean> future = new CompletableFuture<>();
```

Comment:
Instead of completing this future inside the `getUnitsByIds()...` pipeline, you can return the result of `getUnitsByIds()...`.  That will make the method more compact.